Refactor execution package

tiwanari opened the pull request:
I refactored `execution` package and it affected classes in other packages like `RepetitionBreakBlock`.

tiwanari committed:
refactor execution package

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionActivityBase.java:
```diff
@@ -310,7 +305,7 @@ private void onConnectionError() {
                 .show();
 
-        // Inform this activity has been disconnected by the device
+        // inform this device has been disconnected by the device
```
Could you please fix this sentence? I could not understand the meaning.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,28 +17,19 @@
- * @version 1.0 7-July-2013
+ * A container class that has the condition of program execution.
  */
 public class ExecutionCondition {
```
I'm not sure what kind of use cases you expect,
but I do not like these access controls since it does not protect itself.
And this seems provide inconsistent interfaces; while ifStack is public, you also add a kind of setter for that.
In short, why not encapsulate this?

tiwanari replied:
I'm sorry for late replay 😢 

Originally, I prepared `pushSelectionResult(...)` as a helper method. With it, users (mainly the method is used in *Block classes) can use it without considering the program count. But you're right and I will prepare getters and make ifStack private.

mandaiy reviewed (commented):
Hi, I had a quick look at your commit.
To my understanding, you haven't made big changes.
Is my understanding right?
If so, basically it's up to you to approve this PR.

tiwanari committed:
fixed comments on an Activity.

tiwanari committed:
update the appearance of selection blocks

tiwanari committed:
encupsalate ExecutionCondition and fix blocks based on the change

tiwanari commented:
Hi, I fixed codes based on our conversations and simplify the logic of [isThrough()](https://github.com/PileProject/drive/pull/34/files#diff-8a805b762f3f42f7747aa3d3d048eecdR136) method by removing not tested logics for nested ifs. Based on the change, I also changed the appearance of selection blocks like:
![screenshot_20160923-013007](https://cloud.githubusercontent.com/assets/2739915/18757648/c9db667e-812f-11e6-8474-ceafa4f2761b.png)

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,48 +17,186 @@
+        int middle = (block.getRight() + block.getLeft()) / 2;
+
+        // go through this block if (true && left) || (!false && right)
+        return (ifStatus.result && middle <= middleIf) || (!ifStatus.result && middleIf < middle);
```
You should separate the graphics-specific logic from real-programming-related logic, in case you want to revert the block images.

tiwanari replied:
Thanks and what do you think where we should keep it? ExecutionThread was the original location of the long version of this codes.

mandaiy replied:
The best way I came up with is to make a new class that holds graphics-specific-if logic, which might implement an interface common with other strategies.
Or the logic could be in <code>Block</code> class?

I think in future we should separate geometric information from <code>BlockBase</code> class because we are going to implement text-based programming interface in our app.
We can use one <code>BlockBase</code> instance as an atomic operation, which does not need geometric information. When users write some code in our app, then we should parse it, convert it into list of <code>BlockBase</code> instances then we can reuse our code base we already have (though we might need more pretty class name).

So the geometric information should leave from the block class and a new interface, say <code>VisualBlock</code> interface or whatever, should be responsible for the information.
Then the if-logic should be in the interface.

But if we could compile a program and could know next indexes of if-blocks, this conversation would be useless.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,48 +17,186 @@
+     * Set the program count.
+     * @param pc
+     */
+    public void setProgramCount(int pc) {
```
does this method really work?

tiwanari replied:
Sorry and thanks. It does nothing.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,48 +17,186 @@
+    private int mProgramCount;
+    private ArrayList<BlockBase> mBlocks;
+
+    public ExecutionCondition(ArrayList<BlockBase> blocks) {
```
does this class really depend on the ArrayList? If not so, you should replace with just "List"

tiwanari replied:
Finally, I have decided to change it to `List` according to your comments. But let me leave a note for this. The interface of List does not specify the time complexity of `get(int index)` and this program needs random accesses frequently. That's why I have used ArrayList not other List. But the complexity can be ignored because the size of the list is small enough in practical.

mandaiy replied:
Sorry, I missed the use of <code>get</code>. In my opinion you should stick to <code>ArrayList</code>. Or you could convert <code>List</code> into  <code>ArrayList</code> if the list is not <code>instanceof RandomAccess</code>

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionThread.java:
```diff
@@ -158,155 +117,56 @@ private void finalizeExecution() {
         try {
-            Thread.sleep(milli); // Sleep for milli sec
+            Thread.sleep(milli); // sleep for milli sec
         } catch (InterruptedException e) {
```
it would be better to rethrow as non-check exception? same comment applies to other lines

tiwanari replied:
In this case, InterruptedException often occurs while a program execution to stop the program immediately by a user action (e.g., 'stop button'). So, as for this method, the exception can be ignored. But I will fix other methods, thanks :)

mandaiy replied:
that's why I hate Java and checked-exceptions.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionThread.java:
```diff
@@ -25,118 +25,78 @@
-                 mCondition.programCount < mCondition.blocks.size();
-                 mCondition.programCount++) {
-                // Halt execution
+            for (mCondition.setProgramCount(0);
```
do you have any reasons that you did not include initialize logic of program count in the constructor?
(also it could be better to use while statement? but this heavily depends on your taste)

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,48 +17,186 @@
+    private Stack<IfStatus> mIfStack;
+    private int mBeginningOfCurrentLoop;
+    private int mProgramCount;
+    private ArrayList<BlockBase> mBlocks;
```
if the blocks do not change, the field should be final

mandaiy reviewed (changes requested):
Totally this PR contains a lot of solid work, particularly the logic becomes really cleaner in programming processing. I would appreciate if you could fix this according to my comments.

tiwanari committed:
[WIP] update codes based on comments

tiwanari committed:
update classes based on our discussion

tiwanari commented:
I fixed codes based on our discussion and added a logic class for visual block programming.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionThread.java:
```diff
@@ -146,167 +105,62 @@ public void run() {
-            Thread.sleep(milli); // Sleep for milli sec
+            Thread.sleep(milli); // sleep for milli sec
         } catch (InterruptedException e) {
             e.printStackTrace();
```
Is this <code>e.printStackTrace()</code> necessary?

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,48 +17,196 @@
+     * @return
+     * @throws IndexOutOfBoundsException
+     */
+    public BlockBase getBlock(int index) throws IndexOutOfBoundsException {
```
Since users of this class should not have access to blocks by indexing, it's better to hide this method by adding a getter with a selection result instance.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionCondition.java:
```diff
@@ -17,48 +17,196 @@
+     * @param pc program count
+     * @throws IndexOutOfBoundsException
+     */
+    public void setProgramCount(int pc) throws IndexOutOfBoundsException {
```
Stray method; It can be removed. Or do you have particular use cases for this?

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionThread.java:
```diff
@@ -25,119 +25,78 @@
 
-                // Stop temporarily
+                // stop temporarily
                 if (mStop) {
```
It's really painful to understand this if block.
If <code>mController.halt()</code> is idempotent, you could simplify this?

tiwanari replied:
> If mController.halt() is idempotent, you could simplify this?

Yeah, this is true. But `halt()` always sends `stop` commands (\* 2 for left&right) to a robot and it floods buffers. Anyway, I'll try to make it clearer.

mandaiy commented on app/src/main/java/com/pileproject/drive/execution/ExecutionThread.java:
```diff
@@ -146,167 +105,62 @@ public void run() {
         try {
             mController.finalize();
         } catch (RuntimeException e) {
             e.printStackTrace();
```
Do you really need this catch clause and <code>e.printStackTrace()</code>?

mandaiy reviewed (changes requested):
Added some notes. Some of them are bit controversial, so I'm open to a discussion.

tiwanari committed:
update based on comments

tiwanari commented:
Updated codes for check.

tiwanari committed:
add a conversion from Non-RandomAccess List to ArrayList
